### PR TITLE
[sourcekitd] Improve CursorInfo, RelatedIdents, and local rename to better handle VarDecls involved in case fallthroughs

### PR DESCRIPTION
**Pull request by nathawes**
We weren't renaming (via local rename) or returning (RelatedIdents) all occurrences of `x` in the cases like the below:
```
  case .first(let x), .second(let x):
    print("foo \(x)")
    fallthrough
  case .third(let x):
    print("bar \(x)")
```
We would previously only rename/return occurrences within the case statement the query was made in (ignoring `fallthrough`s) and for cases with multiple `CaseLabelItem`s (as in the first case above) we would only rename/return the occurrence in the first one. CursorInfo similarly returned nothing when invoked on a variable in all but first first case label item of each case.

These changes will also allow us to handle the implicit `VarDecl`s that @gottesmm plans to introduce into case statement bodies soon.

### CODE REVIEW

**Comment by nathawes** [comment]
@swift-ci test

File: test/SourceKit/CursorInfo/cursor_vardecl_across_fallthrough.swift [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+// CHECK1DECL: source.lang.swift.decl.var.local (13:19-13:20)
+// CHECK1REF: source.lang.swift.ref.var.local (13:19-13:20)
+
+// RUN: %sourcekitd-test -req=cursor -pos=16:20 %s -- %mcp_opt %s | %FileCheck -check-prefixes=CHECKX,CHECK2DECL %s
+// RUN: %sourcekitd-test -req=cursor -pos=16:42 %s -- %mcp_opt %s | %FileCheck -check-prefixes=CHECKX,CHECK2DECL2 %s
+// RUN: %sourcekitd-test -req=cursor -pos=17:18 %s -- %mcp_opt %s | %FileCheck -check-prefixes=CHECKX,CHECK2REF %s
+
+// CHECK2DECL: source.lang.swift.decl.var.local (16:20-16:21)
+// CHECK2DECL2: source.lang.swift.decl.var.local (16:42-16:43)
+// CHECK2REF: source.lang.swift.ref.var.local (16:20-16:21)
```
Comment (nathawes) [inline test/SourceKit/CursorInfo/cursor_vardecl_across_fallthrough.swift]: @gottesmm This is the location jump-to-definition will land on when invoked on the reference on line 17. Once you have your changes in I expect the `(16:20-16:21)` on this line to change to `(16:42-16:43)`. I.e. cursor info should say the second 'let x' on line 16 is the one being referred to.

File: test/SourceKit/CursorInfo/cursor_vardecl_across_fallthrough.swift [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+// CHECK3DECL: source.lang.swift.decl.var.local (13:26-13:27)
+// CHECK3REF: source.lang.swift.ref.var.local (13:26-13:27)
+
+// RUN: %sourcekitd-test -req=cursor -pos=16:27 %s -- %mcp_opt %s | %FileCheck -check-prefixes=CHECKY,CHECK4DECL %s
+// RUN: %sourcekitd-test -req=cursor -pos=16:49 %s -- %mcp_opt %s | %FileCheck -check-prefixes=CHECKY,CHECK4DECL2 %s
+// RUN: %sourcekitd-test -req=cursor -pos=17:23 %s -- %mcp_opt %s | %FileCheck -check-prefixes=CHECKY,CHECK4REF %s
+
+// CHECK4DECL: source.lang.swift.decl.var.local (16:27-16:28)
+// CHECK4DECL2: source.lang.swift.decl.var.local (16:49-16:50)
+// CHECK4REF: source.lang.swift.ref.var.local (16:27-16:28)
```
Comment (nathawes) [inline test/SourceKit/CursorInfo/cursor_vardecl_across_fallthrough.swift]: Same here. I expect this to change to `(16:49-16:50)` with @gottesmm's changes.

**Comment by nathawes** [comment]
@swift-ci please smoke test

**Review by akyrtzi — APPROVED** [review]
LGTM, make sure to get it in swift-5.1 branch as well.

**Comment by gottesmm** [comment]
@akyrtzi @nathawes This is the cherry-pick PR that you guys need before this can be cherry-picked. https://github.com/apple/swift/pull/23449

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
I think these test statuses are out of date b/c of the jenkins master fun times yesterday. I restarted the test.

**Pull request merged**